### ISSUE TITLE
Where is the edge collapse operation in the code?

### ISSUE THREAD

**Issue by liangqianqian123** [ranahanocka/MeshCNN#44]
![1](https://user-images.githubusercontent.com/33091152/69618174-67168000-1074-11ea-83af-efce120234f5.PNG)

I want to know where is the change of edge feature after edge pooling in the code? Is it in the file mesh_pool.py? I only find the function "def __pool_side(self, mesh, edge_id, mask, edge_groups, side)", but this function didn't involve the change of the edge feature.
Thank you please.

**Comment by ranahanocka** [comment]
Hi @liangqianqian123 ,

It is called a "mesh union" in the code. Specifically, see [union_groups line](https://github.com/ranahanocka/MeshCNN/blob/29c30a4d6dee123cacbb662e3f5666420c0e94c5/models/layers/mesh_pool.py#L107) in `__pool_side`. This is defined in the [Mesh_union](https://github.com/ranahanocka/MeshCNN/blob/master/models/layers/mesh_union.py) class.

**Comment by liangqianqian123** [comment]
I want to know where is the change of edge feature, such as average operation of the edge features. Is it in the class MeshUnion "rebuild_features_average"?


`class MeshUnion:
    def __init__(self, n, device=torch.device('cpu')):
        self.__size = n
        self.rebuild_features = self.rebuild_features_average
        self.groups = torch.eye(n).to(device)

    def union(self, source, target):
        self.groups[target, :] += self.groups[source, :]
      
    def remove_group(self, index):
        return

    def get_group(self, edge_key):
        return self.groups[edge_key, :]

    def get_occurrences(self):
        return torch.sum(self.groups, 0)

    def get_groups(self, tensor_mask):
        self.groups = torch.clamp(self.groups, 0, 1)
        return self.groups[tensor_mask, :]

    def rebuild_features_average(self, features, mask, target_edges):
        self.prepare_groups(features, mask)
        fe = torch.matmul(features.squeeze(-1), self.groups)
        occurrences = torch.sum(self.groups, 0).expand(fe.shape)
        fe = fe / occurrences
        padding_b = target_edges - fe.shape[1]
        if padding_b > 0:
            padding_b = ConstantPad2d((0, padding_b, 0, 0), 0)
            fe = padding_b(fe)
        return fe

    def prepare_groups(self, features, mask):
        tensor_mask = torch.from_numpy(mask)
        #tensor to numpy
        #   a.numpy()
        # numpy to tensor
        #   torch.from_numpy(a)
        self.groups = torch.clamp(self.groups[tensor_mask, :], 0, 1).transpose_(1, 0)
        padding_a = features.shape[1] - self.groups.shape[0]
        if padding_a > 0:
            padding_a = ConstantPad2d((0, 0, 0, padding_a), 0)
            self.groups = padding_a(self.groups)`

**Comment by ranahanocka** [comment]
> Is it in the class MeshUnion "rebuild_features_average"?

yep!
